dash/include/dash/map/UnorderedMap.h

```diff
@@ -643,6 +611,68 @@ class UnorderedMap
     return result;
   }
 
+  iterator insert(
+    const_iterator hint,
+    const value_type & value)
+  {
+    Timer::timestamp_t ts_enter = Timer::Now(), ts_insert, ts_find, d_insert, d_find;
+    auto key = value.first;
+    auto mapped = value.second;
+
+    DASH_ASSERT(_globmem != nullptr);
+    DASH_LOG_DEBUG("UnorderedMap.insert()", "key:", key, "mapped:", mapped);
+
+    auto unit = _key_hash(key);
+
+    iterator found = _end;
+
+    if (_myid == unit) {
+      DASH_LOG_TRACE("UnorderedMap.insert", "local element key lookup");
+
+      ts_find = Timer::Now();
+      auto lbegin = static_cast<value_type *>(_lbegin);
+      auto lend = static_cast<value_type *>(_lend);
+      const_local_iterator liter = std::find_if(
+                  _lbegin, _lend,
+                   [&](const value_type & v) {
+                     return _key_equal(v.first, key);
+                   });
+      d_find = Timer::ElapsedSince(ts_find);
+
+      if (liter != _lend) {
+        found = iterator(this, _myid, liter.pos());
+      }
+
+    } else  {
+      DASH_LOG_TRACE("UnorderedMap.insert", "element key lookup");
+      iterator found = find(key);
+    }
+    DASH_LOG_TRACE_VAR("UnorderedMap.insert", found);
+
+    iterator res;
+    if (found != _end) {
+      DASH_LOG_TRACE("UnorderedMap.insert", "key found");
+      // Existing element found, no insertion:
+      res = found;
+    } else {
+      DASH_LOG_TRACE("UnorderedMap.insert", "key not found");
+      // Unit mapped to the new element's key by the hash function:
+      DASH_LOG_TRACE("UnorderedMap.insert", "target unit:", unit);
+      // No element with specified key exists, insert new value.
+      ts_insert = Timer::Now();
```

Comment:
Benchmarking / debugging leftovers? If these changes are intended to improve performance, add a benchmark example.